File: app/src/main/java/com/app/cancellingretrofitrequests/java/ui/MainActivity.java
package com.app.cancellingretrofitrequests.java.ui;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.lifecycle.Observer;

import com.app.cancellingretrofitrequests.R;
import com.app.cancellingretrofitrequests.java.model.GenericDataModel;
import com.app.cancellingretrofitrequests.java.model.SearchResultModel;
import com.app.cancellingretrofitrequests.java.repository.DataRepository;

public class MainActivity extends AppCompatActivity {

    private AppCompatEditText mSearchEditText;
    private AppCompatTextView mResultTextView;
    private ProgressBar mProgressLoading;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
    }

    private void initView() {
        mSearchEditText = findViewById(R.id.searchEditText);
        mResultTextView = findViewById(R.id.resultTextView);
        mProgressLoading = findViewById(R.id.progressLoading);

        mSearchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                startSearching(s.toString());
            }
        });
    }

    private void startSearching(String searchQuery) {
        DataRepository.getSearchResults(searchQuery).observe(this, new Observer<GenericDataModel<SearchResultModel>>() {
            @Override
            public void onChanged(GenericDataModel<SearchResultModel> searchResultModelGenericDataModel) {
                checkAndSetData(searchResultModelGenericDataModel);
            }
        });
    }

    private void checkAndSetData(GenericDataModel<SearchResultModel> searchResultModelGenericDataModel) {
        if (searchResultModelGenericDataModel != null) {
            if (searchResultModelGenericDataModel.isLoading != null && searchResultModelGenericDataModel.isLoading) {
                mResultTextView.setVisibility(View.GONE);
                mProgressLoading.setVisibility(View.VISIBLE);
            } else {
                if (searchResultModelGenericDataModel.isSuccess != null && searchResultModelGenericDataModel.isSuccess) {
                    SearchResultModel data = searchResultModelGenericDataModel.data;
                    if (data != null && data.getStatus().equalsIgnoreCase("ok") &&
                            data.getData() != null && data.getData().getItems() != null &&
                            data.getData().getItems().size() > 0) {
                        mResultTextView.setText(data.toString());
                    } else {
                        mResultTextView.setText("No data");
                    }
                } else {
                    mResultTextView.setText("No data");
                }
                mResultTextView.setVisibility(View.VISIBLE);
                mProgressLoading.setVisibility(View.GONE);
            }
        } else {
            mResultTextView.setVisibility(View.GONE);
            mProgressLoading.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/app/cancellingretrofitrequests/kotlin/ui/MainActivity.kt
package com.app.cancellingretrofitrequests.kotlin.ui

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.app.cancellingretrofitrequests.R
import com.app.cancellingretrofitrequests.kotlin.model.GenericDataModel
import com.app.cancellingretrofitrequests.kotlin.model.SearchResultModel
import com.app.cancellingretrofitrequests.kotlin.repository.DataRepository
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity() {

    private var apiJob: Job? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        searchEditText?.addTextChangedListener(object: TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                resultTextView?.visibility = View.GONE
                progressLoading?.visibility = View.VISIBLE
                apiJob?.cancel()

                startSearching(s.toString())
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }
        })
    }

    private fun startSearching(searchQuery: String?) {
        apiJob = CoroutineScope(Dispatchers.IO).launch {
            DataRepository.getSearchResults(searchQuery)

            withContext(Dispatchers.Main) {
                DataRepository.searchResultsLiveData.observe(
                    this@MainActivity,
                    Observer { genericDataModel: GenericDataModel<SearchResultModel>? ->
                        run {
                            checkAndSetData(genericDataModel)
                        }
                    })
            }
        }
    }

    private fun checkAndSetData(genericDataModel: GenericDataModel<SearchResultModel>?) {
        if (genericDataModel != null) {
            if (genericDataModel.isLoading == true) {
                resultTextView?.visibility = View.GONE
                progressLoading?.visibility = View.VISIBLE
            } else {
                if (genericDataModel.isSuccess == true) {
                    val data = genericDataModel.data
                    if (data?.status.equals("OK", true) && data?.data?.items?.size ?: 0 > 0) {
                        resultTextView?.text = data.toString()
                    } else {
                        resultTextView?.text = "No data"
                    }
                } else {
                    resultTextView?.text = "No data"
                }

                resultTextView?.visibility = View.VISIBLE
                progressLoading?.visibility = View.GONE
            }
        } else {
            resultTextView?.visibility = View.GONE
            progressLoading?.visibility = View.GONE
        }
    }
}
